xarray/core/utils.py

```diff
@@ -435,6 +435,31 @@ def FrozenDict(*args, **kwargs) -> Frozen:
     return Frozen(dict(*args, **kwargs))
 
 
+class HybridMappingProxy(Mapping[K, V]):
+    """Implements the Mapping interface. Uses the wrapped mapping for item lookup
+    and a separate wrapped keys collection for iteration.
+
+    Can be used to construct a mapping object from another dict-like object without
+    eagerly accessing its items or when a mapping object is expected but only
+    iteration over keys is actually used.
```

Comment:
Could we note that it's the caller's responsibility to ensure that `keys` are the keys of `mapping`, i.e. that this doesn't validate that?

Reply:
That is a good suggestion. I will add a note.